manta-crypto/src/eclair/cmp.rs

```diff
@@ -56,30 +60,155 @@ where
+                compiler.assert_eq(lhs, rhs);
+            }
+        }
+    }
+}
+
+impl<T, Rhs, COM> PartialEq<Box<[Rhs]>, COM> for Box<[T]>
+where
+    COM: Has<bool>,
+    Bool<COM>: Constant<COM, Type = bool> + BitAnd<Bool<COM>, COM, Output = Bool<COM>>,
+    T: PartialEq<Rhs, COM>,
+{
+    #[inline]
+    fn eq(&self, rhs: &Box<[Rhs]>, compiler: &mut COM) -> Bool<COM> {
+        if self.len() != rhs.len() {
+            false.as_constant(compiler)
+        } else {
+            let mut are_equal = true.as_constant::<Bool<COM>>(compiler);
+            for (lhs, rhs) in self.iter().zip(rhs.iter()) {
+                are_equal = are_equal.bitand(lhs.eq(rhs, compiler), compiler);
+            }
+            are_equal
+        }
+    }
+
+    #[inline]
+    fn assert_equal(&self, rhs: &Box<[Rhs]>, compiler: &mut COM)
+    where
+        COM: Assert,
+    {
+        if self.len() != rhs.len() {
+            let not_equal = false.as_constant(compiler);
+            compiler.assert(&not_equal);
+        } else {
+            for (lhs, rhs) in self.iter().zip(rhs.iter()) {
+                compiler.assert_eq(lhs, rhs);
+            }
+        }
+    }
+}
+
+impl<T, Rhs, const N: usize, COM> PartialEq<Array<Rhs, N>, COM> for Array<T, N>
+where
+    COM: Has<bool>,
+    Bool<COM>: Constant<COM, Type = bool> + BitAnd<Bool<COM>, COM, Output = Bool<COM>>,
+    T: PartialEq<Rhs, COM>,
+{
+    #[inline]
+    fn eq(&self, rhs: &Array<Rhs, N>, compiler: &mut COM) -> Bool<COM> {
+        let mut are_equal = true.as_constant::<Bool<COM>>(compiler);
+        for (lhs, rhs) in self.iter().zip(rhs) {
+            are_equal = are_equal.bitand(lhs.eq(rhs, compiler), compiler);
+        }
+        are_equal
+    }
+
+    #[inline]
+    fn assert_equal(&self, rhs: &Array<Rhs, N>, compiler: &mut COM)
+    where
+        COM: Assert,
+    {
+        for (lhs, rhs) in self.iter().zip(rhs) {
+            compiler.assert_eq(lhs, rhs);
+        }
+    }
+}
+
+impl<T, Rhs, const N: usize, COM> PartialEq<BoxArray<Rhs, N>, COM> for BoxArray<T, N>
 where
-    T: cmp::PartialEq<Rhs>,
+    COM: Has<bool>,
+    Bool<COM>: Constant<COM, Type = bool> + BitAnd<Bool<COM>, COM, Output = Bool<COM>>,
+    T: PartialEq<Rhs, COM>,
 {
     #[inline]
-    fn eq(&self, rhs: &Rhs, _: &mut ()) -> bool {
-        self.eq(rhs)
+    fn eq(&self, rhs: &BoxArray<Rhs, N>, compiler: &mut COM) -> Bool<COM> {
+        let mut are_equal = true.as_constant::<Bool<COM>>(compiler);
+        for (lhs, rhs) in self.iter().zip(rhs) {
+            are_equal = are_equal.bitand(lhs.eq(rhs, compiler), compiler);
+        }
+        are_equal
     }
 
     #[inline]
-    fn ne(&self, rhs: &Rhs, _: &mut ()) -> bool {
-        self.ne(rhs)
+    fn assert_equal(&self, rhs: &BoxArray<Rhs, N>, compiler: &mut COM)
+    where
+        COM: Assert,
+    {
+        for (lhs, rhs) in self.iter().zip(rhs) {
+            compiler.assert_eq(lhs, rhs);
+        }
     }
 }
-*/
```

Comment:
General comment on these implementations of PartialEq: it feels that they might be particular cases of a more general implementation, e.g., something like T : IntoIterator with T::Item : PartialEq.

Reply:
We can't naively do that, but there is some compression we can do. Let's move to an issue: https://github.com/Manta-Network/manta-rs/issues/207.